File: android/iden3coreapi/src/main/java/com/iden3/iden3coreapi/Iden3Identity.kt
package com.iden3.iden3coreapi

import iden3mobile.*
import java.util.ArrayList
import java.util.HashMap

class Iden3Identity(private var identity: Identity) {

    /**
     * Stop an Iden3 identity.
     *
     */
    fun stopIdentity() {
        identity.stop()
    }

    /**
     * Request a claim from an issuer.
     *
     * @param issuerUrl: Url of the issuer.
     * @param data: data needed by the issuer for requesting a claim.
     * @param ticketListener: Listener of tickets received after requesting the claim.
     *
     * @return An Iden3Ticket.
     */
    @Throws(Exception::class)
    fun requestClaim(issuerUrl: String, data: String, ticketListener: Iden3TicketListener?) : Iden3Ticket? {
        if (ticketListener == null) {
            return Iden3Ticket(identity.requestClaim(issuerUrl, data))
        } else {
            identity.requestClaimWithCb(issuerUrl, data) {
                ticket, exception ->
                if (exception != null) {
                    ticketListener.onError(exception)
                } else if (ticket != null) {
                    ticketListener.onTicketReceived(Iden3Ticket(ticket))
                }
            }
        }
        return null
    }

    /**
     * Prove a claim from an verifier.
     *
     * @param verifierUrl: url of the verifier.
     * @param credentialId: credential id needed by the verifier for proving a claim.
     * @param withZKProof: Boolean that indicates if zero-knowledge proof shoud be used or not.
     * @param proveClaimListener: Listener of proofs received after proving a claim.
     *
     * @return Boolean that indicates if the claim has been proven or not.
     */
    @Throws(Exception::class)
    fun proveClaim(verifierUrl: String, credentialId: String, withZKProof: Boolean, proveClaimListener: Iden3ProveClaimListener?) : Boolean?  {
        if (withZKProof) {
            if (proveClaimListener == null) {
                return identity.proveClaim(verifierUrl, credentialId)
            } else {
                identity.proveClaimWithCb(verifierUrl, credentialId) {
                    isProven, exception ->
                    if (exception != null) {
                        proveClaimListener.onError(exception)
                    } else {
                        proveClaimListener.onClaimProofReceived(isProven)
                    }
                }
            }
        } else {
            if (proveClaimListener == null) {
                return identity.proveClaimZK(verifierUrl, credentialId)
            } else {
                identity.proveClaimZKWithCb(verifierUrl, credentialId) {
                        isProven, exception ->
                        if (exception != null) {
                            proveClaimListener.onError(exception)
                        } else {
                            proveClaimListener.onClaimProofReceived(isProven)
                        }
                }
            }
        }
        return null
    }

    /**
     * List of claims of the identity.
     *
     * @return ArrayList with the claims of the identity.
     */
    @Throws(Exception::class)
    fun listClaims() : ArrayList<HashMap<*,*>> {
        val claims = ArrayList<HashMap<*, *>>()
        try {
            val cdb = identity.claimDB
            cdb.iterateClaimsJSON { key, claim ->
                val map = HashMap<String, Any?>()
                map["key"] = key
                map["claim"] = claim
                claims.add(map)
            }
        } catch (e: java.lang.Exception) {

        }
        return claims
    }

    /**
     * List of credentials of the identity.
     *
     * @return ArrayList with the credentials of the identity.
     */
    @Throws(Exception::class)
    fun listCredentials() : ArrayList<HashMap<*,*>> {
        val credentials = ArrayList<HashMap<*,*>>()
        try {
            val cdb = identity.claimDB
            cdb.iterateCredExistJSON { key, credential ->
                val map = HashMap<String, Any?>()
                map["key"] = key
                map["credential"] = credential
                credentials.add(map)
            }
        } catch (e: Exception) {

        }
        return credentials
    }

    /**
     * List of events of the identity.
     *
     * @return ArrayList with the events of the identity.
     */
    fun listEvents() : ArrayList<Iden3Ticket> {
        val events = ArrayList<Iden3Ticket>()
        identity.tickets.iterate {
            events.add(Iden3Ticket(it))
        }
        return events
    }

    /**
     * Cancel an event of the identity.
     *
     * @return Boolean that indicates if the event has been cancelled or not.
     */
    fun cancelEvent(eventId: String) : Boolean {
        return try {
            identity.tickets.cancelTicket(eventId)
            true
        } catch (e: Exception) {
            false
        }
    }
}

File: ios/iden3CoreSDK/iden3CoreSDK/Iden3Identity.swift
//
//  Iden3Identity.swift
//  iden3CoreSDK
//
//  Created by Iden3 on 12/06/2020.
//  Copyright © 2020 Iden3. All rights reserved.
//

import Foundation
import Iden3mobile

public final class Iden3Identity {
    
    // MARK: - Properties
    
    private var identity: Iden3mobileIdentity
    
    // MARK: - Initializers
       
    init(identity: Iden3mobileIdentity) {
        self.identity = identity
    }
    
    /// Stops the Iden3 identity.
    ///
    public func stopIdentity() {
        identity.stop()
    }
    
    /// Request a claim from an issuer.
    ///
    /// - parameters:
    ///   - issuerUrl: Url of the issuer
    ///   - data: data needed by the issuer for requesting a claim.
    ///   - ticketDelegate: Delegate of tickets received after requesting the claim.
    ///
    ///   return Iden3Ticket.
    ///
    public func requestClaim(issuerUrl: String, data: String, ticketDelegate: Iden3TicketDelegate?) throws -> Iden3Ticket? {
        if (ticketDelegate == nil) {
            do {
                return Iden3Ticket(try identity.requestClaim(issuerUrl, data: data))
            } catch {
                return nil
            }
        } else {
            identity.requestClaim(withCb: issuerUrl, data: data, c: nil)
        }
        return nil
    }
    
    /// Sends a credentialValidity build from the given credentialExistance to a verifier.
    ///
    /// - parameters:
    ///   - verifierUrl: url of the verifier
    ///   - credentialId: credential id needed by the verifier for proving a claim.
    ///   - withZKProof: Boolean that indicates if zero-knowledge proof shoud be used or not.
    ///   - proveClaimDelegate: Delegate of proofs received after proving a claim.
    ///
    ///   return Bool should be true if the verifier accepted the prove as valid
    ///
    public func proveClaim(verifierUrl: String, credentialId: String, withZKProof: Bool, proveClaimDelegate: Iden3ProveClaimDelegate?) throws -> Bool {
        if (withZKProof) {
            if (proveClaimDelegate == nil) {
                do {
                    var result: ObjCBool = false
                    try identity.proveClaim(verifierUrl, credID: credentialId, ret0_:&result)
                    if (result.boolValue == true) {
                        return true
                    } else {
                        return false
                    }
                } catch {
                    return false
                }
            } else {
                identity.proveClaim(withCb: verifierUrl, credID: credentialId, c: nil)
            }
        } else {
            if (proveClaimDelegate == nil) {
                do {
                    var result: ObjCBool = false
                    try identity.proveClaimZK(verifierUrl, credID: credentialId, ret0_:&result)
                    if (result.boolValue == true) {
                        return true
                    } else {
                        return false
                    }
                } catch {
                    return false
                }
            } else {
                identity.proveClaimZK(withCb: verifierUrl, credID: credentialId, c: nil)
            }
        }
        return false
    }
    
    /// List of claims of the identity.
    ///
    /// return ArrayList with the claims of the identity.
    ///
    public func listClaims() -> Array<Dictionary<String, String>> {
        let claims = [Dictionary<String,String>]()
        let cdb = identity.claimDB
        do {
            let callback = ClaimsIterator<Iden3mobileClaimDBIterFnerProtocol>(claims)
            try cdb?.iterateClaimsJSON(callback as? Iden3mobileClaimDBIterFnerProtocol)
        } catch {

        }
        return claims
    }
    
    /// List of credentials of the identity.
    ///
    /// return ArrayList with the claims of the identity.
    ///
    public func listCredentials() -> Array<Dictionary<String, String>> {
        let credentials = [Dictionary<String,String>]()
        let cdb = identity.claimDB
        do {
            let callback = ClaimsIterator<Iden3mobileClaimDBIterFnerProtocol>(credentials)
            try cdb?.iterateCredExistJSON(callback as? Iden3mobileClaimDBIterFnerProtocol)
        } catch {

        }
        return credentials
    }
    
    private class ClaimsIterator<Iden3mobileClaimDBIterFnerProtocol> {
           
        private var claims : Array<Dictionary<String, String>>
           
        init(_ claims: Array<Dictionary<String, String>>) {
            self.claims = claims
        }
           
        func fn(_ key: String?, claim: String?, ret0_: UnsafeMutablePointer<ObjCBool>?) throws {
            var map = Dictionary<String, String>()
            map["key"] = key
            map["claim"] = claim
            self.claims.append(map)
        }
    }

    /**
     * List of events of the identity.
     *
     * @return ArrayList with the events of the identity.
     */
    public func listEvents() -> Array<Iden3Ticket> {
        let events = [Iden3Ticket]()
        do {
        try identity.tickets?.iterate(TicketIterator<Iden3mobileTicketOperatorProtocol>(events) as? Iden3mobileTicketOperatorProtocol)
        } catch {
            
        }
        return events
    }
    
    private class TicketIterator<Iden3mobileTicketOperatorProtocol> {
           
        private var tickets : Array<Iden3Ticket>
           
        init(_ tickets: Array<Iden3Ticket>) {
            self.tickets = tickets
        }
        
        func iterate(_ ticket: Iden3mobileTicket?, ret0_: UnsafeMutablePointer<ObjCBool>?) throws {
            if (ticket != nil) {
                self.tickets.append(Iden3Ticket(ticket!))
            }
        }
    }

    /**
     * Cancel an event of the identity.
     *
     * @return Boolean that indicates if the event has been cancelled or not.
     */
    public func cancelEvent(eventId: String) -> Bool {
        do {
            try identity.tickets?.cancelTicket(eventId)
            return true
        } catch  {
            return false
        }
    }
}
